juju/application.py

```diff
@@ -706,19 +701,23 @@ async def refresh(
         if parsed_url.schema is None:
             raise JujuError(f'A ch: or cs: schema is required for application refresh, given : {str(parsed_url)}')
 
-        if revision is not None:
-            origin.revision = revision
+        current_origin = charm_url_origin_result.charm_origin
+        if path is not None or (switch is not None and is_local_charm(switch)):
+            await self.local_refresh(current_origin, force, force_series,
+                                     force_units, path or switch, resources)
+            return
 
-        # Make the source-specific changes to the origin/channel/url
-        # (and also get the resources necessary to deploy the (destination) charm -- for later)
-        origin.source = Source.CHARM_HUB.value
-        if channel:
-            ch = Channel.parse(channel).normalize()
-            origin.risk = ch.risk
-            origin.track = ch.track
+        if channel is not None:
+            channel = Channel.parse(channel).normalize()
 
-        charmhub = self.model.charmhub
-        charm_resources = await charmhub.list_resources(charm_name)
+        origin = client.CharmOrigin(
```

Comment:
let's move this out into a helper function and have some unit tests against this to protect the logic for any regressions.

Reply:
What logic exactly are you referring to here?

Reply:
More like decisions I meant to say, we make a bunch of decisions here, it would be nice if we have unit tests that assert those decisions so if any of them are violated in the future we'd catch it before it breaks.

```
            source=str(Source.CHARM_HUB),
            risk=channel.risk if channel else current_origin.risk,
            track=channel.track if channel else current_origin.track,
            revision=revision or current_origin.revision,
            base=current_origin.base,
            architecture=current_origin.get('architecture', DEFAULT_ARCHITECTURE),
```

Reply:
Done